{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **METK Barley SNP-Chip:** Exploring the correlation between barleyâ€™s genetic makeup and its protein content\n",
    "## Importing and editing the necessary datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "snip_data = pd.read_csv(\"SNIP_DATA.csv\")\n",
    "barley_data = pd.read_csv(\"BARLEY_DATA.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Giving the first column in \"SNIP_DATA.csv\" a header since it was originally unnamed\n",
    "snip_data.rename(columns={snip_data.columns[0]: 'SNP'}, inplace=True)\n",
    "\n",
    "# Dropping ID column for \"BARLEY_DATA.csv\"\n",
    "barley_data = barley_data.drop(columns='Id')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cleaning dataset 'SNIP_DATA.csv' based on the following criteria: \n",
    "### 1. Handling 'failed' values by replacing them with NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "snip_data.replace('failed', np.nan, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Removing SNPs with the same allele across all varieties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Excluding the first column\n",
    "snp_columns = snip_data.columns[1:]\n",
    "\n",
    "# Filtering rows where there's more than one unique value in the SNP columns \n",
    "# This ignores NaN values so that if the row is [\"A\", \"A\", \"A\", \"A\", NaN, NaN] then it is still removed although there's 2 different values\n",
    "snip_data = snip_data[snip_data[snp_columns].apply(lambda row: row.dropna().nunique() > 1, axis=1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Removing barley varieties that are not present in both datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to normalize variety names because\n",
    "# in dataset barley_data some varieties presented in form '5777.7.1.2' are written as 5777712 in snip_data\n",
    "def normalize_variety_names(variety):\n",
    "    return str(variety).replace('.', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalizing variety names in barley_data\n",
    "barley_data['Nimi'] = barley_data['Nimi'].apply(normalize_variety_names)\n",
    "\n",
    "# Extracting variety names from snip_data (columns starting from the second column)\n",
    "snip_varieties = set(snip_data.columns[1:])\n",
    "\n",
    "# Extracting variety names from barley_data (row values in the appropriate column)\n",
    "barley_varieties = set(barley_data['Nimi'])\n",
    "\n",
    "# Finding common varieties\n",
    "common_varieties = snip_varieties.intersection(barley_varieties)\n",
    "\n",
    "# Filtering snip_data to keep only common varieties\n",
    "snip_data = snip_data[['SNP'] + list(common_varieties)]\n",
    "\n",
    "# Filtering barley_data to keep only rows with common varieties\n",
    "barley_data = barley_data[barley_data['Nimi'].isin(common_varieties)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing the datasets \n",
    "### **In preparation for finding correlations between protein content and genetic makeup**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating a copy with only the barley variety and protein columns\n",
    "protein_data = barley_data[['Nimi', 'Proteiin']].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging datasets on variety name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transposing snip_data to have barley varieties as rows not columns\n",
    "snip_data_transposed = snip_data.set_index('SNP').T.reset_index()\n",
    "snip_data_transposed.rename(columns={'index': 'Nimi'}, inplace=True)\n",
    "\n",
    "\n",
    "# Merging protein_data with the transposed snip_data\n",
    "merged_data = protein_data.merge(snip_data_transposed, on='Nimi', how='inner')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Seeing what different types of values we have as alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G    562756\n",
      "A    540192\n",
      "C    492895\n",
      "T    429012\n",
      "R      2030\n",
      "Y      1405\n",
      "K       466\n",
      "M       416\n",
      "S        82\n",
      "W        45\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Extracting allele columns\n",
    "allele_columns = merged_data.columns[2:]\n",
    "\n",
    "# Flattening all values from allele columns into a single series and dropping NaN\n",
    "all_alleles = merged_data[allele_columns].stack().dropna()\n",
    "\n",
    "# Counting the occurrences of each allele value\n",
    "allele_counts = all_alleles.value_counts()\n",
    "\n",
    "print(allele_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Brief biological explanation:**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| Nucleotide Symbol | Full Name                       |\n",
    "|-------------------|---------------------------------|\n",
    "| A                 | Adenine                         |\n",
    "| C                 | Cytosine                        |\n",
    "| G                 | Guanine                         |\n",
    "| T                 | Thymine                         |\n",
    "| R                 | Guanine / Adenine (purine)      |\n",
    "| Y                 | Cytosine / Thymine (pyrimidine) |\n",
    "| K                 | Guanine / Thymine               |\n",
    "| M                 | Adenine / Cytosine              |\n",
    "| S                 | Guanine / Cytosine              |\n",
    "| W                 | Adenine / Thymine               |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating numeric values for alleles and imputing NaN values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Nimi</th>\n",
       "      <th>Proteiin</th>\n",
       "      <th>BK_01</th>\n",
       "      <th>BK_03</th>\n",
       "      <th>BK_05</th>\n",
       "      <th>BK_08</th>\n",
       "      <th>BK_10</th>\n",
       "      <th>BK_12</th>\n",
       "      <th>BK_14</th>\n",
       "      <th>BK_17</th>\n",
       "      <th>...</th>\n",
       "      <th>TGBA15K-TG0384</th>\n",
       "      <th>TGBA15K-TG0385</th>\n",
       "      <th>TGBA15K-TG0386</th>\n",
       "      <th>TGBA15K-TG0388</th>\n",
       "      <th>TGBA15K-TG0395</th>\n",
       "      <th>TGBA15K-TG0400</th>\n",
       "      <th>TGBA15K-TG0402_NC_MA</th>\n",
       "      <th>TGBA15K-TG0402_NG_MA</th>\n",
       "      <th>TGBA15K-TG0403</th>\n",
       "      <th>TGBA15K-TG0409</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Amidala</td>\n",
       "      <td>12.2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Amy</td>\n",
       "      <td>12.7</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Anneli</td>\n",
       "      <td>13.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Anni</td>\n",
       "      <td>12.4</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Annika</td>\n",
       "      <td>11.2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>6006142</td>\n",
       "      <td>12.1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>6006153</td>\n",
       "      <td>11.3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>6011421</td>\n",
       "      <td>11.4</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>6012243</td>\n",
       "      <td>11.6</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>6012252</td>\n",
       "      <td>12.4</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>171 rows Ã— 11898 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Nimi  Proteiin  BK_01  BK_03  BK_05  BK_08  BK_10  BK_12  BK_14  \\\n",
       "0    Amidala      12.2    3.0    3.0    1.0    2.0    0.6    0.0    0.0   \n",
       "1        Amy      12.7    3.0    3.0    1.0    2.0    0.0    0.0    0.0   \n",
       "2     Anneli      13.0    3.0    3.0    3.0    2.0    0.0    0.0    0.0   \n",
       "3       Anni      12.4    3.0    3.0    1.0    1.0    0.0    0.0    0.0   \n",
       "4     Annika      11.2    3.0    3.0    1.0    2.0    0.0    0.0    0.0   \n",
       "..       ...       ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "166  6006142      12.1    3.0    3.0    1.0    1.0    0.0    0.0    0.0   \n",
       "167  6006153      11.3    3.0    3.0    1.0    2.0    0.0    0.0    0.0   \n",
       "168  6011421      11.4    3.0    3.0    1.0    1.0    0.0    0.0    0.0   \n",
       "169  6012243      11.6    3.0    3.0    1.0    2.0    0.0    0.0    0.0   \n",
       "170  6012252      12.4    3.0    3.0    1.0    2.0    0.0    0.0    0.0   \n",
       "\n",
       "     BK_17  ...  TGBA15K-TG0384  TGBA15K-TG0385  TGBA15K-TG0386  \\\n",
       "0      1.0  ...             1.0             3.0             2.0   \n",
       "1      1.0  ...             1.0             3.0             2.0   \n",
       "2      1.0  ...             1.0             2.0             2.0   \n",
       "3      2.0  ...             1.0             3.0             3.0   \n",
       "4      1.0  ...             1.0             3.0             2.0   \n",
       "..     ...  ...             ...             ...             ...   \n",
       "166    1.0  ...             1.0             3.0             2.0   \n",
       "167    1.0  ...             1.0             3.0             2.0   \n",
       "168    1.0  ...             1.0             3.0             3.0   \n",
       "169    1.0  ...             1.0             3.0             3.0   \n",
       "170    1.0  ...             1.0             3.0             3.0   \n",
       "\n",
       "     TGBA15K-TG0388  TGBA15K-TG0395  TGBA15K-TG0400  TGBA15K-TG0402_NC_MA  \\\n",
       "0               2.0             1.0             0.0                   3.0   \n",
       "1               2.0             1.0             0.0                   3.0   \n",
       "2               2.0             1.0             0.0                   3.0   \n",
       "3               2.0             1.0             0.0                   3.0   \n",
       "4               2.0             1.0             0.0                   3.0   \n",
       "..              ...             ...             ...                   ...   \n",
       "166             2.0             1.0             0.0                   3.0   \n",
       "167             2.0             1.0             0.0                   3.0   \n",
       "168             2.0             1.0             0.0                   3.0   \n",
       "169             2.0             1.0             0.0                   3.0   \n",
       "170             2.0             1.0             0.0                   3.0   \n",
       "\n",
       "     TGBA15K-TG0402_NG_MA  TGBA15K-TG0403  TGBA15K-TG0409  \n",
       "0                     3.0             1.0             2.0  \n",
       "1                     3.0             1.0             2.0  \n",
       "2                     3.0             1.0             0.0  \n",
       "3                     3.0             1.0             2.0  \n",
       "4                     3.0             1.0             2.0  \n",
       "..                    ...             ...             ...  \n",
       "166                   3.0             1.0             2.0  \n",
       "167                   3.0             1.0             2.0  \n",
       "168                   3.0             1.0             2.0  \n",
       "169                   3.0             1.0             2.0  \n",
       "170                   3.0             1.0             2.0  \n",
       "\n",
       "[171 rows x 11898 columns]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.impute import KNNImputer\n",
    "\n",
    "\n",
    "# Encoding non-numeric data: assigning integer values to each allele type\n",
    "allele_mapping = {\n",
    "    'A': 0,\n",
    "    'C': 1,\n",
    "    'G': 2,\n",
    "    'T': 3,\n",
    "    'R': 4,\n",
    "    'Y': 5,\n",
    "    'K': 6,\n",
    "    'M': 7,\n",
    "    'S': 8,\n",
    "    'W': 9\n",
    "}\n",
    "\n",
    "# Apply encoding to SNP columns\n",
    "columns_to_impute = merged_data.drop(['Nimi', 'Proteiin'], axis=1)\n",
    "encoded_columns = columns_to_impute.map(lambda x: allele_mapping.get(x, np.nan))  \n",
    "\n",
    "# creating a data frame from the list\n",
    "Before_imputation = pd.DataFrame(encoded_columns)\n",
    "\n",
    "# create an object for KNNImputer\n",
    "imputer = KNNImputer(n_neighbors=5)\n",
    "After_imputation = imputer.fit_transform(Before_imputation)\n",
    "\n",
    "# Convert the result back to a DataFrame with the original column names\n",
    "After_imputation_df = pd.DataFrame(After_imputation, columns=columns_to_impute.columns)\n",
    "\n",
    "# Add the 'Nimi' and 'Proteiin' columns back to the imputed dataset\n",
    "merged_data = pd.concat([merged_data[['Nimi', 'Proteiin']], After_imputation_df], axis=1)\n",
    "\n",
    "# Display the final imputed dataset\n",
    "merged_data\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Performing one-hot encoding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Empty DataFrame to store one-hot encoded SNP data\n",
    "encoded_snps = pd.DataFrame(index=merged_data.index)\n",
    "\n",
    "# Looping through SNP columns and apply one-hot encoding\n",
    "for column in merged_data.columns[2:]:  # Skip 'Nimi' and 'Proteiin'\n",
    "    one_hot = pd.get_dummies(merged_data[column], prefix=column, dtype=bool)\n",
    "    encoded_snps = pd.concat([encoded_snps, one_hot], axis=1)\n",
    "\n",
    "# Concatenating back with 'Nimi' and 'Proteiin' columns\n",
    "encoded_data = pd.concat([merged_data[['Nimi', 'Proteiin']], encoded_snps], axis=1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each SNP is represented as multiple binary columns, one for each **observed** allele. For instance, if a SNP site had the alleles \"A,\" \"C,\" and \"T,\" it would be encoded as three separate columns (e.g., SNP_A, SNP_C, SNP_T), where a value of True signifies the presence of that allele, and False indicates its absence.\n",
    "\n",
    "**Let's look at our encoded dataset:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Nimi</th>\n",
       "      <th>Proteiin</th>\n",
       "      <th>BK_01_2.0</th>\n",
       "      <th>BK_01_2.8</th>\n",
       "      <th>BK_01_3.0</th>\n",
       "      <th>BK_03_1.0</th>\n",
       "      <th>BK_03_3.0</th>\n",
       "      <th>BK_05_1.0</th>\n",
       "      <th>BK_05_3.0</th>\n",
       "      <th>BK_05_5.0</th>\n",
       "      <th>...</th>\n",
       "      <th>TGBA15K-TG0400_2.0</th>\n",
       "      <th>TGBA15K-TG0402_NC_MA_2.0</th>\n",
       "      <th>TGBA15K-TG0402_NC_MA_3.0</th>\n",
       "      <th>TGBA15K-TG0402_NG_MA_2.0</th>\n",
       "      <th>TGBA15K-TG0402_NG_MA_3.0</th>\n",
       "      <th>TGBA15K-TG0403_0.0</th>\n",
       "      <th>TGBA15K-TG0403_1.0</th>\n",
       "      <th>TGBA15K-TG0409_0.0</th>\n",
       "      <th>TGBA15K-TG0409_2.0</th>\n",
       "      <th>TGBA15K-TG0409_4.0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Amidala</td>\n",
       "      <td>12.2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Amy</td>\n",
       "      <td>12.7</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Anneli</td>\n",
       "      <td>13.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Anni</td>\n",
       "      <td>12.4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Annika</td>\n",
       "      <td>11.2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>6006142</td>\n",
       "      <td>12.1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>6006153</td>\n",
       "      <td>11.3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>6011421</td>\n",
       "      <td>11.4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>6012243</td>\n",
       "      <td>11.6</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>6012252</td>\n",
       "      <td>12.4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>171 rows Ã— 26709 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Nimi  Proteiin  BK_01_2.0  BK_01_2.8  BK_01_3.0  BK_03_1.0  BK_03_3.0  \\\n",
       "0    Amidala      12.2      False      False       True      False       True   \n",
       "1        Amy      12.7      False      False       True      False       True   \n",
       "2     Anneli      13.0      False      False       True      False       True   \n",
       "3       Anni      12.4      False      False       True      False       True   \n",
       "4     Annika      11.2      False      False       True      False       True   \n",
       "..       ...       ...        ...        ...        ...        ...        ...   \n",
       "166  6006142      12.1      False      False       True      False       True   \n",
       "167  6006153      11.3      False      False       True      False       True   \n",
       "168  6011421      11.4      False      False       True      False       True   \n",
       "169  6012243      11.6      False      False       True      False       True   \n",
       "170  6012252      12.4      False      False       True      False       True   \n",
       "\n",
       "     BK_05_1.0  BK_05_3.0  BK_05_5.0  ...  TGBA15K-TG0400_2.0  \\\n",
       "0         True      False      False  ...               False   \n",
       "1         True      False      False  ...               False   \n",
       "2        False       True      False  ...               False   \n",
       "3         True      False      False  ...               False   \n",
       "4         True      False      False  ...               False   \n",
       "..         ...        ...        ...  ...                 ...   \n",
       "166       True      False      False  ...               False   \n",
       "167       True      False      False  ...               False   \n",
       "168       True      False      False  ...               False   \n",
       "169       True      False      False  ...               False   \n",
       "170       True      False      False  ...               False   \n",
       "\n",
       "     TGBA15K-TG0402_NC_MA_2.0  TGBA15K-TG0402_NC_MA_3.0  \\\n",
       "0                       False                      True   \n",
       "1                       False                      True   \n",
       "2                       False                      True   \n",
       "3                       False                      True   \n",
       "4                       False                      True   \n",
       "..                        ...                       ...   \n",
       "166                     False                      True   \n",
       "167                     False                      True   \n",
       "168                     False                      True   \n",
       "169                     False                      True   \n",
       "170                     False                      True   \n",
       "\n",
       "     TGBA15K-TG0402_NG_MA_2.0  TGBA15K-TG0402_NG_MA_3.0  TGBA15K-TG0403_0.0  \\\n",
       "0                       False                      True               False   \n",
       "1                       False                      True               False   \n",
       "2                       False                      True               False   \n",
       "3                       False                      True               False   \n",
       "4                       False                      True               False   \n",
       "..                        ...                       ...                 ...   \n",
       "166                     False                      True               False   \n",
       "167                     False                      True               False   \n",
       "168                     False                      True               False   \n",
       "169                     False                      True               False   \n",
       "170                     False                      True               False   \n",
       "\n",
       "     TGBA15K-TG0403_1.0  TGBA15K-TG0409_0.0  TGBA15K-TG0409_2.0  \\\n",
       "0                  True               False                True   \n",
       "1                  True               False                True   \n",
       "2                  True                True               False   \n",
       "3                  True               False                True   \n",
       "4                  True               False                True   \n",
       "..                  ...                 ...                 ...   \n",
       "166                True               False                True   \n",
       "167                True               False                True   \n",
       "168                True               False                True   \n",
       "169                True               False                True   \n",
       "170                True               False                True   \n",
       "\n",
       "     TGBA15K-TG0409_4.0  \n",
       "0                 False  \n",
       "1                 False  \n",
       "2                 False  \n",
       "3                 False  \n",
       "4                 False  \n",
       "..                  ...  \n",
       "166               False  \n",
       "167               False  \n",
       "168               False  \n",
       "169               False  \n",
       "170               False  \n",
       "\n",
       "[171 rows x 26709 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "encoded_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finding correlations"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
