{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **METK Barley SNP-Chip:** Exploring the correlation between barleyâ€™s genetic makeup and its protein content\n",
    "## Importing and editing the necessary datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "snip_data = pd.read_csv(\"SNIP_DATA.csv\")\n",
    "barley_data = pd.read_csv(\"BARLEY_DATA.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Giving the first column in \"SNIP_DATA.csv\" a header since it was originally unnamed\n",
    "snip_data.rename(columns={snip_data.columns[0]: 'SNP'}, inplace=True)\n",
    "\n",
    "# Dropping ID column for \"BARLEY_DATA.csv\"\n",
    "barley_data = barley_data.drop(columns='Id')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cleaning dataset 'SNIP_DATA.csv' based on the following criteria: \n",
    "### 1. Handling 'failed' values by replacing them with NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "snip_data.replace('failed', pd.NA, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Removing SNPs with the same allele across all varieties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Excluding the first column\n",
    "snp_columns = snip_data.columns[1:]\n",
    "\n",
    "# Filtering rows where there's more than one unique value in the SNP columns \n",
    "# This ignores NaN values so that if the row is [\"A\", \"A\", \"A\", \"A\", NaN, NaN] then it is still removed although there's 2 different values\n",
    "snip_data = snip_data[snip_data[snp_columns].apply(lambda row: row.dropna().nunique() > 1, axis=1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Removing barley varieties that are not present in both datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to normalize variety names because\n",
    "# in dataset barley_data some varieties presented in form '5777.7.1.2' are written as 5777712 in snip_data\n",
    "def normalize_variety_names(variety):\n",
    "    return str(variety).replace('.', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalizing variety names in barley_data\n",
    "barley_data['Nimi'] = barley_data['Nimi'].apply(normalize_variety_names)\n",
    "\n",
    "# Extracting variety names from snip_data (columns starting from the second column)\n",
    "snip_varieties = set(snip_data.columns[1:])\n",
    "\n",
    "# Extracting variety names from barley_data (row values in the appropriate column)\n",
    "barley_varieties = set(barley_data['Nimi'])\n",
    "\n",
    "# Finding common varieties\n",
    "common_varieties = snip_varieties.intersection(barley_varieties)\n",
    "\n",
    "# Filtering snip_data to keep only common varieties\n",
    "snip_data = snip_data[['SNP'] + list(common_varieties)]\n",
    "\n",
    "# Filtering barley_data to keep only rows with common varieties\n",
    "barley_data = barley_data[barley_data['Nimi'].isin(common_varieties)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing the datasets \n",
    "### **In preparation for finding correlations between protein content and genetic makeup**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating a copy with only the barley variety and protein columns\n",
    "protein_data = barley_data[['Nimi', 'Proteiin']].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging datasets on variety name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Nimi</th>\n",
       "      <th>Proteiin</th>\n",
       "      <th>BK_01</th>\n",
       "      <th>BK_03</th>\n",
       "      <th>BK_05</th>\n",
       "      <th>BK_08</th>\n",
       "      <th>BK_10</th>\n",
       "      <th>BK_12</th>\n",
       "      <th>BK_14</th>\n",
       "      <th>BK_17</th>\n",
       "      <th>...</th>\n",
       "      <th>TGBA15K-TG0384</th>\n",
       "      <th>TGBA15K-TG0385</th>\n",
       "      <th>TGBA15K-TG0386</th>\n",
       "      <th>TGBA15K-TG0388</th>\n",
       "      <th>TGBA15K-TG0395</th>\n",
       "      <th>TGBA15K-TG0400</th>\n",
       "      <th>TGBA15K-TG0402_NC_MA</th>\n",
       "      <th>TGBA15K-TG0402_NG_MA</th>\n",
       "      <th>TGBA15K-TG0403</th>\n",
       "      <th>TGBA15K-TG0409</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Amidala</td>\n",
       "      <td>12.2</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Amy</td>\n",
       "      <td>12.7</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Anneli</td>\n",
       "      <td>13.0</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Anni</td>\n",
       "      <td>12.4</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Annika</td>\n",
       "      <td>11.2</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>6006142</td>\n",
       "      <td>12.1</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>6006153</td>\n",
       "      <td>11.3</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>6011421</td>\n",
       "      <td>11.4</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>6012243</td>\n",
       "      <td>11.6</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>6012252</td>\n",
       "      <td>12.4</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>171 rows Ã— 11898 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Nimi  Proteiin BK_01 BK_03 BK_05 BK_08 BK_10 BK_12 BK_14 BK_17  ...  \\\n",
       "0    Amidala      12.2  <NA>     T     C     G  <NA>     A     A     C  ...   \n",
       "1        Amy      12.7     T     T     C     G     A     A     A     C  ...   \n",
       "2     Anneli      13.0     T     T     T     G     A     A     A     C  ...   \n",
       "3       Anni      12.4     T     T     C     C     A     A     A     G  ...   \n",
       "4     Annika      11.2     T     T     C     G     A     A     A     C  ...   \n",
       "..       ...       ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   \n",
       "166  6006142      12.1     T     T     C     C     A     A     A     C  ...   \n",
       "167  6006153      11.3     T     T     C     G     A     A     A     C  ...   \n",
       "168  6011421      11.4     T     T     C     C     A     A     A     C  ...   \n",
       "169  6012243      11.6     T     T     C     G     A     A     A     C  ...   \n",
       "170  6012252      12.4     T     T     C     G     A     A     A     C  ...   \n",
       "\n",
       "    TGBA15K-TG0384 TGBA15K-TG0385 TGBA15K-TG0386 TGBA15K-TG0388  \\\n",
       "0                C              T              G              G   \n",
       "1                C              T              G              G   \n",
       "2                C              G              G              G   \n",
       "3                C              T              T              G   \n",
       "4                C              T              G              G   \n",
       "..             ...            ...            ...            ...   \n",
       "166              C              T              G              G   \n",
       "167              C              T              G              G   \n",
       "168              C              T              T              G   \n",
       "169              C              T              T              G   \n",
       "170              C              T              T              G   \n",
       "\n",
       "    TGBA15K-TG0395 TGBA15K-TG0400 TGBA15K-TG0402_NC_MA TGBA15K-TG0402_NG_MA  \\\n",
       "0                C              A                    T                    T   \n",
       "1                C              A                    T                    T   \n",
       "2                C              A                    T                    T   \n",
       "3                C              A                    T                    T   \n",
       "4                C              A                    T                    T   \n",
       "..             ...            ...                  ...                  ...   \n",
       "166              C              A                    T                    T   \n",
       "167              C              A                    T                    T   \n",
       "168              C              A                    T                    T   \n",
       "169              C              A                    T                    T   \n",
       "170              C              A                    T                    T   \n",
       "\n",
       "    TGBA15K-TG0403 TGBA15K-TG0409  \n",
       "0                C              G  \n",
       "1                C              G  \n",
       "2                C              A  \n",
       "3                C              G  \n",
       "4                C              G  \n",
       "..             ...            ...  \n",
       "166              C              G  \n",
       "167              C              G  \n",
       "168              C              G  \n",
       "169              C              G  \n",
       "170              C              G  \n",
       "\n",
       "[171 rows x 11898 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Transposing snip_data to have barley varieties as rows not columns\n",
    "snip_data_transposed = snip_data.set_index('SNP').T.reset_index()\n",
    "snip_data_transposed.rename(columns={'index': 'Nimi'}, inplace=True)\n",
    "\n",
    "\n",
    "# Merging protein_data with the transposed snip_data\n",
    "merged_data = protein_data.merge(snip_data_transposed, on='Nimi', how='inner')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Seeing what different types of values we have as alleles in preparation for a one-hot encoding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G    562756\n",
      "A    540192\n",
      "C    492895\n",
      "T    429012\n",
      "R      2030\n",
      "Y      1405\n",
      "K       466\n",
      "M       416\n",
      "S        82\n",
      "W        45\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Extracting allele columns\n",
    "allele_columns = merged_data.columns[2:]\n",
    "\n",
    "# Flattening all values from allele columns into a single series and dropping NaN\n",
    "all_alleles = merged_data[allele_columns].stack().dropna()\n",
    "\n",
    "# Counting the occurrences of each allele value\n",
    "allele_counts = all_alleles.value_counts()\n",
    "\n",
    "# Displaying the counts\n",
    "print(allele_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Brief biological explanation:**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| Nucleotide Symbol | Full Name                       |\n",
    "|-------------------|---------------------------------|\n",
    "| A                 | Adenine                         |\n",
    "| C                 | Cytosine                        |\n",
    "| G                 | Guanine                         |\n",
    "| T                 | Thymine                         |\n",
    "| R                 | Guanine / Adenine (purine)      |\n",
    "| Y                 | Cytosine / Thymine (pyrimidine) |\n",
    "| K                 | Guanine / Thymine               |\n",
    "| M                 | Adenine / Cytosine              |\n",
    "| S                 | Guanine / Cytosine              |\n",
    "| W                 | Adenine / Thymine               |\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
